flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/BatchingStateChangeStore.java

```diff
@@ -0,0 +1,173 @@
+import static java.util.concurrent.TimeUnit.SECONDS;
+import static org.apache.flink.util.ExceptionUtils.findThrowable;
+import static org.apache.flink.util.Preconditions.checkState;
+
+/**
+ * A {@link StateChangeStore} that waits for some configured amount of time before passing the
+ * accumulated state changes to the actual store. The writes are executed asynchronously.
+ */
+@ThreadSafe
+class BatchingStateChangeStore implements StateChangeStore {
+    private static final Logger LOG = LoggerFactory.getLogger(BatchingStateChangeStore.class);
+
+    private final StateChangeStore delegate;
+    private final ScheduledExecutorService scheduler;
+    private final long scheduleDelayMs;
+    private final long sizeThresholdBytes;
+
+    @GuardedBy("scheduled")
+    private final Queue<StoreTask> scheduled;
+
+    @GuardedBy("scheduled")
+    private long scheduledSizeInBytes;
+
+    @GuardedBy("scheduled")
+    private Future<?> scheduledFuture = CompletableFuture.completedFuture(null);
+
+    private AtomicReference<Throwable> error = new AtomicReference<>(null);
+
+    BatchingStateChangeStore(
+            long persistDelayMs, long sizeThresholdBytes, StateChangeStore delegate) {
+        this(
+                persistDelayMs,
+                sizeThresholdBytes,
+                delegate,
+                SchedulerFactory.create(1, "ChangelogRetryScheduler", LOG));
+    }
+
+    BatchingStateChangeStore(
+            long persistDelayMs,
+            long sizeThresholdBytes,
+            StateChangeStore delegate,
+            ScheduledExecutorService scheduler) {
+        this.scheduleDelayMs = persistDelayMs;
+        this.scheduled = new LinkedList<>();
+        this.scheduler = scheduler;
+        this.sizeThresholdBytes = sizeThresholdBytes;
+        this.delegate = delegate;
+    }
+
+    @Override
+    public void save(StoreTask storeTask) {
+        Collection<StateChangeSet> changeSets = storeTask.changeSets;
+        if (error.get() != null) {
+            LOG.debug("don't persist {} changesets, already failed", changeSets.size());
+            storeTask.fail(error.get());
+            return;
+        }
+        LOG.debug("persist {} changeSets", changeSets.size());
+        try {
+            synchronized (scheduled) {
+                scheduled.add(storeTask);
+                scheduledSizeInBytes += storeTask.getSize();
+                scheduleUploadIfNeeded();
+            }
+        } catch (Exception e) {
+            storeTask.fail(e);
+            throw e;
+        }
+    }
+
+    private void scheduleUploadIfNeeded() {
+        checkState(holdsLock(scheduled));
+        if (scheduledFuture.isDone() || isOverSizeThresholdAndCancellationSucceded()) {
+            scheduleUpload();
+        }
+    }
+
+    private boolean isOverSizeThresholdAndCancellationSucceded() {
+        return scheduledSizeInBytes >= sizeThresholdBytes && scheduledFuture.cancel(false);
+    }
```

Comment:
I think the 2nd check may result in threshold being exceeded but not scheduled for upload:
1. T1: add 1 byte, schedule future 1
2. Executor: waits and eventually runs future 1, calls `scheduleUploadIfNeeded` and returns without scheduling; now thread gets suspended by the OS (so future 1 isn't completed)
2. T1: add more bytes exceeding the threshold; try to cancel future 1 and fail => future2 is not scheduled by T1
4. Executor also doesn't schedule any future 

WDYT?

Reply:
IIUC, if `drainAndSave` is being executing but it already passed the re-scheduling section (2) then nothing will be scheduled.